modules/alloy/src/main/ts/ephox/alloy/positioning/view/PositionCss.ts

```diff
@@ -23,16 +26,45 @@ const NuPositionCss = (
   bottom
 });
 
-const applyPositionCss = (element: SugarElement, position: PositionCss): void => {
+const applyPositionCss = (element: SugarElement, position: PositionCss, decision: Optional<RepositionDecision>): void => {
   const addPx = (num: number) => num + 'px';
 
-  Css.setOptions(element, {
+  const cssOptions = {
     position: Optional.some(position.position),
     left: position.left.map(addPx),
     top: position.top.map(addPx),
     right: position.right.map(addPx),
     bottom: position.bottom.map(addPx)
-  });
+  };
+
+  const changedFromTopToBottom = isElementTopAligned(element) && isDecisionBottomAligned(decision);
+  const changedFromBottomToTop = isElementBottomAligned(element) && isDecisionTopAligned(decision);
+
+  if (changedFromTopToBottom || changedFromBottomToTop) {
+    Css.set(element, 'position', 'absolute');
+
+    const getValue = (key: 'top' | 'left' | 'bottom' | 'right') => {
+      if (cssOptions[key].isSome()) {
+        return Optional.some(Css.get(element, key));
+      } else {
+        return Optional.none<string>();
+      }
+    };
+
+    const intermediateCssOptions = {
+      position: cssOptions.position,
+      top: getValue('top'),
+      right: getValue('right'),
+      bottom: getValue('bottom'),
+      left: getValue('left'),
+    };
+
+    Css.setOptions(element, intermediateCssOptions);
+    Class.add(element, contextBarTransitionClass);
```

Comment:
Update: I found where it's cleaned up. Is there any way we can make sure (even if it's just by comments) that the element passed into this function comes from a component that has the right event handling in place?

Also, as the adding the class logic and the removing the class logic are in completely different places, we should probably add a test that walks through the entire lifecycle (creating the toolbar, pinning it to one end, changing it to be pinned at the other end, waiting for the transition to complete) that ensures that the right classes are on the element at each stage of the process. Otherwise, logic like this that's spread over multiple files is really easy to remove accidentally if you can't see the context for why it's there (which is at the other end of alloy).

Reply:
I don't think any other elements can reach this code, since it requires Layout-elements, so it shouldn't be anything else, though we can add a comment about it right there to try and be extra clear.
And a comment where it's removed.

As for the test, a good idea, but it'd be good if it could be made into its own issue. Tight deadline for this, and I would need some time to figure out how to make the test.

Reply:
Since the class is now provided as an optional, this should be safe, right?